File: inst/python/movmean.py
import numpy as np


def movmean_py_np(x, half_win=1):
    '''
    moving mean in numpy
            
    x = np.array([1,4,5,9,2,3,4])
    movmean_python(x,2)   #函数检验   

    By: Li Chenxi
    '''
    m = half_win
    win = 2*m+1
    n = len(x)
    y = np.zeros(n)
    for i in range(0, n):
        if i < m:
            ind = np.arange(i+m+1)
            y[i] = sum(x[ind])/len(x[ind])
        elif i >= m and i <= (n-m-1):
            ind = np.arange((i-m), (i+m+1))
            y[i] = sum(x[ind])/win
        else:
            ind = np.arange((i-m), n)
            y[i] = sum(x[ind])/len(x[ind])
    return y


def movmean_py(x, halfwin=1):
    """
    moving mean in Python array

    By: 赵雨欣子
    """
    m = halfwin
    win = m*2 + 1
    n = len(x)
    y = np.zeros(n)
    for i in range(n):
        if m <= i < (n-m):
          y[i] = sum(x[(i-m):(i+m+1)])/win
        if (i < m):
          y[i] = sum(x[0:(m+i+1)])/len(x[0:(m+i+1)])
        if i >= n-m:
          y[i] = sum(x[(i-m):n])/len(x[(i-m):n])
    return y

# if __name__ == "__main__":
#     movmean()
# x = np.array([1, 2, 3, 4])
# x = [1, 2, 3, 4]
# print(x[1])

File: inst/julia/movmean.jl
using Statistics

function movmean(x::AbstractArray{Union{Missing, T}, 1},  halfwin::Integer = 2) where {T <: Real}
    # NA and Inf will automatically ignored
    n = length(x)
    z = zeros(n)
    
    # is_valid = isfinite.(x) .& (.!ismissing.(x))
    @inbounds for i = 1:n
        i_begin = i <= halfwin ? 1 : i - halfwin
        i_end = i <= n - halfwin ? i + halfwin : n
        
        val = 0.0; len = 0;
        @inbounds for j in i_begin:i_end
            if !ismissing(x[j]) & isfinite(x[j])
                val += x[j];
                len += 1;
            end
        end 
        # ind_raw = (i_begin:i_end)
        # ind = ind_raw[is_valid[ind_raw]]
        if len > 0
            z[i] = val/len
        else
            z[i] = missing
        end
    end
    z
end

function movmean(x::AbstractArray{T, 1}, halfwin::Integer = 2) where {T <: Real}
    n = length(x)
    z = zeros(n)
    for i = 1:n
        i_begin = i <= halfwin ? 1 : i - halfwin
        i_end = i <= n - halfwin ? i + halfwin : n
        # println((i_begin, i_end))
        z[i] = mean(@view x[i_begin:i_end])
    end
    z
end

weightedMean(x::AbstractArray{T, 1}, w::AbstractArray{T2, 1}) where {T <: Real, T2 <: Real} = sum(x .* w) / sum(w)

# 4 times slower
function movmean(x::AbstractArray{T, 1}, w::AbstractArray{T2, 1}, halfwin::Integer = 2) where {T <: Real, T2 <: Real}
    n = length(x)
    z = zeros(n)
    for i = 1:n
        i_begin = i <= halfwin ? 1 : i - halfwin
        i_end = i <= n - halfwin ? i + halfwin : n

        # println((i_begin, i_end))
        z[i] = weightedMean(view(x, i_begin:i_end), view(w, i_begin:i_end))
        # z[i] = weightedMean(x[i_begin:i_end], w[i_begin:i_end])
        # z[i] = sum(view(x, i_begin:i_end) .* view(w, i_begin:i_end)) / sum(view(w, i_begin:i_end))
    end
    z
end


export movmean, weightedMean

# using RCall
# R"""
# x = c(NA, 1, 2, 3, Inf)
# """
# @rget x
# movmean(x)
